### PR TITLE
Port funcall

### PR DESCRIPTION
**Pull request by harryfei**

### CODE REVIEW

**Comment by shaleh** [comment]
I think this diff brings the code closer to the original C intent. What do you think Harry?

```
diff --git a/rust_src/src/eval.rs b/rust_src/src/eval.rs
index ddee851cdb..eedf094645 100644
--- a/rust_src/src/eval.rs
+++ b/rust_src/src/eval.rs
@@ -997,9 +997,9 @@ pub fn funcall(args: &mut [LispObject]) -> LispObject {

     let mut val = Qnil;

-    loop {
-        let original_fun = args[0];
+    let original_fun = args[0];

+    loop {
         if original_fun.is_nil() {
             xsignal!(Qvoid_function, original_fun);
         }
@@ -1028,8 +1028,12 @@ pub fn funcall(args: &mut [LispObject]) -> LispObject {
             };
         } else if unsafe { COMPILEDP(fun) || MODULE_FUNCTIONP(fun) } {
             val = unsafe { funcall_lambda(fun, numargs as isize, args[1..].as_mut_ptr()) };
-        } else if fun.is_cons() {
-            let funcar = fun.as_cons_or_error().car();
+        } else {
+            let funcar = if let Some(cons) = fun.as_cons() {
+                cons.car()
+            } else {
+                xsignal!(Qinvalid_function, original_fun);
+            };

             if !funcar.is_symbol() {
                 xsignal!(Qinvalid_function, original_fun);
@@ -1046,8 +1050,6 @@ pub fn funcall(args: &mut [LispObject]) -> LispObject {
             } else {
                 xsignal!(Qinvalid_function, original_fun);
             }
-        } else {
-            xsignal!(Qinvalid_function, original_fun);
         }

         break;
@@ -1065,7 +1067,7 @@ pub fn funcall(args: &mut [LispObject]) -> LispObject {
         current_thread.m_specpdl_ptr = current_thread.m_specpdl_ptr.offset(-1);
     }

-    return val;
+    val
 }

 include!(concat!(env!("OUT_DIR"), "/eval_exports.rs"));
```

**Comment by harryfei** [comment]
@yeah, you are right.

Actually, your diff is very close to my first version of the porting. 

However, I made some changes to the first version because I think the new version is more readable. Besides, we will find that the intent of the new version is not changed if we think about it more deeply.

If you think our porting should be more similar to the C counterpart, I will modify this PR. :smile:

File: rust_src/src/eval.rs [hunk 1]
```diff
@@ -948,4 +951,118 @@ fn run_hook_with_args_internal(
...
+        let fun = if original_fun.is_symbol() {
+            original_fun.as_symbol_or_error().get_indirect_function()
+        } else {
+            original_fun
+        };
+
+        if fun.is_nil() {
+            xsignal!(Qvoid_function, original_fun);
+        }
+
```
Comment (shaleh) [inline rust_src/src/eval.rs]: you can use the `if Some()` here too.

File: rust_src/src/eval.rs [hunk 2]
```diff
@@ -948,4 +951,118 @@ fn run_hook_with_args_internal(
...
+            }
+
+            if funcar.eq(Qlambda) || funcar.eq(Qclosure) {
+                val = unsafe { funcall_lambda(fun, numargs as isize, args[1..].as_mut_ptr()) };
+            } else if funcar.eq(Qautoload) {
+                Fautoload_do_load(fun, original_fun, Qnil);
+                unsafe {
+                    check_cons_list();
+                }
+                continue;
```
Comment (shaleh) [inline rust_src/src/eval.rs]: We have the same else twice. How do we debug this if it fails? The C code is cleaner. Can we simplify the Rust?

Comment (harryfei) [reply rust_src/src/eval.rs]: Is this cleaner?

```rust
    if let Some(mut f) = fun.as_subr() {
        val = unsafe { funcall_subr(f.as_mut(), numargs as isize, args[1..].as_mut_ptr()) };
    } else if unsafe { COMPILEDP(fun) || MODULE_FUNCTIONP(fun) } {
        val = unsafe { funcall_lambda(fun, numargs as isize, args[1..].as_mut_ptr()) };
    } else if let Some(cons) = fun.as_cons() {
        let funcar = cons.car();

        if !funcar.is_symbol() {
            xsignal!(Qinvalid_function, original_fun);
        }

        if !funcar.eq(Qlambda) && !funcar.eq(Qclosure) && !funcar.eq(Qautoload) {
            xsignal!(Qinvalid_function, original_fun);
        }

        if funcar.eq(Qautoload) {
            Fautoload_do_load(fun, original_fun, Qnil);
            unsafe { check_cons_list() };
            continue;
        }

        val = unsafe { funcall_lambda(fun, numargs as isize, args[1..].as_mut_ptr()) };
    } else {
        xsignal!(Qinvalid_function, original_fun);
    }

```

File: rust_src/src/eval.rs (hunk 1 above)
Comment (harryfei) [reply rust_src/src/eval.rs]: Fixed.

File: rust_src/src/eval.rs [hunk 3]
```diff
@@ -948,4 +951,118 @@ fn run_hook_with_args_internal(
...
+
+    let mut val = Qnil;
+
+    let original_fun = args[0];
+
+    loop {
+        if original_fun.is_nil() {
+            xsignal!(Qvoid_function, original_fun);
+        }
+
```
[the remark this answers is not in the document]
Comment (harryfei) [reply rust_src/src/eval.rs]: Fixed

File: rust_src/src/eval.rs (hunk 2 above)
Comment (harryfei) [reply rust_src/src/eval.rs]: Reimplemented in a new way.

File: rust_src/src/eval.rs [hunk 4]
```diff
@@ -948,4 +952,143 @@ fn run_hook_with_args_internal(
...
+            return Ok(LispFun::LambdaFun(fun));
+        }
+
+        if let Some(cons) = fun.as_cons() {
+            let funcar = cons.car();
+
+            if !funcar.is_symbol() {
+                return Err(LispFunError::InvalidFun);
+            }
+
```
Comment (shaleh) [inline rust_src/src/eval.rs]: This does not read as nicely as the C code.

Why not use positive checks here?

```
if Qlambda or Qclosure {
    return Ok(LambdaFun);
}

if Qautoload {
   Fautoload_do_load(fun, original_fun, Qnil);
   unsafe { check_cons_list() };
   continue;
}
return Err(LispFunError::InvalidFun);
```

File: rust_src/src/eval.rs [hunk 5]
```diff
@@ -948,4 +952,143 @@ fn run_hook_with_args_internal(
...
+            if current_thread.m_lisp_eval_depth > globals.max_lisp_eval_depth {
+                error!("Lisp nesting exceeds `max-lisp-eval-depth'");
+            }
+        }
+    }
+
+    // The first element in args is the called function.
+    let numargs = args.len() as isize - 1;
+
```
[the remark this answers is not in the document]
Comment (harryfei) [reply rust_src/src/eval.rs]: Fixed

File: rust_src/src/eval.rs [hunk 6]
```diff
@@ -948,4 +952,143 @@ fn run_hook_with_args_internal(
...
+    let original_fun = fun;
+
+    loop {
+        if original_fun.is_nil() {
+            return Err(LispFunError::VoidFun);
+        }
+
+        // Optimize for no indirection.
+        let fun = original_fun
+            .as_symbol()
```
[the remark this answers is not in the document]
Comment (harryfei) [reply rust_src/src/eval.rs]: You are right. Fixed.

File: rust_src/src/eval.rs (hunk 4 above)
Comment (harryfei) [reply rust_src/src/eval.rs]: Fixed

**Comment by harryfei** [comment]
Change: rebase master instead of merge.

File: rust_src/src/eval.rs [hunk 7]
```diff
@@ -565,7 +569,7 @@ pub fn eval(form: LispObject, lexical: LispObject) -> LispObject {
 #[no_mangle]
 pub extern "C" fn apply1(mut func: LispObject, arg: LispObject) -> LispObject {
     if arg == Qnil {
-        unsafe { Ffuncall(1, &mut func) }
+        Ffuncall(1, &mut func)
```
Comment (shanavas786) [inline rust_src/src/eval.rs]: can't we use `funcall` itself here ?

Comment (db48x) [reply rust_src/src/eval.rs]: see `eval.c:1912`. `Ffuncall` is the C name of the lisp function `funcall`, and it's not just a simple wrapper around a single C function. Instead it handles the eval depth check, backtraces, GC, interactive debugging, error handling, the `specpdl_ptr`, and finally it delegates to any of several other function depending on what you're calling (`funcall_subr`, `funcall_lambda`, and `Fautoload_do_load`). Eventually someone will port it to Rust and the `Ffuncall` name will go away.

Comment (harryfei) [reply rust_src/src/eval.rs]: Yeah, I missed that.

Comment (shaleh) [reply rust_src/src/eval.rs]: Good catch @shanavas786

**Pull request merged**